File: work/dp_g.py
# -*- coding: utf-8 -*-
from collections import deque

n,m = map(int, input().split())
edge = [set() for _ in [0]*n]
iny = [0 for _ in [0]*n]
for _ in range(m):
    x,y = map(int, input().split())
    x -= 1
    y -= 1
    edge[x].add(y)
    iny[y] += 1

# from random import randint
# n = 10**1
# m = 10**1
# edge = [set() for _ in [0]*n]
# for _ in range(m):
#     x = randint(0,n-1)
#     y = randint(0,n-1)
#     while x==y:
#         x = randint(0,n-1)
#         y = randint(0,n-1)
#
#     edge[x].add(y)

res = 0
d = [0]*n
for i in [j for j in range(n) if iny[j]==0]:
    if d[i]>0:
        continue
    st = deque([i])
    while len(st)>0:
        s = st.popleft()
        for v in edge[s]:
            if d[v]<d[s]+1:
                d[v] = d[s]+1
                st.append(v)

print(max(d))

File: work/dp_g.cpp
#include<iostream>
#include<vector>
#include<queue>
using namespace std;

int main(){
  int n,m;
  cin>>n>>m;

  vector<vector<int>> edge(n);
  vector<int> iny = vector<int>(n,0);
  for(int i=0;i<m;i++){
    int x,y;
    cin>>x>>y;
    x -= 1;
    y -= 1;
    edge[x].push_back(y);
    iny[y]++;
  }

  vector<int> d = vector<int>(n, 0);
  queue<int> st;
  for(int i=0;i<n;i++){
    if(iny[i]==0){
      st.push(i);
    }
  }

  while(st.size()>0){
    int s = st.front();
    st.pop();
    for(int j=0;j<edge[s].size();j++){
      int v = edge[s][j];
      if(iny[v]>0){
        iny[v]--;
      }
      if(iny[v]==0){
        d[v] = max(d[v], d[s]+1);
        st.push(v);
      }
    }
  }

  int res = 0;
  for(int i=0;i<d.size();i++){
    res = max(res, d[i]);
  }
  cout<<res<<endl;
  return 0;
}
